data-prepper-plugins/lambda-sink/src/main/java/org/opensearch/dataprepper/plugins/sink/lambda/LambdaSinkService.java

```diff
@@ -0,0 +1,237 @@
+                        codec.start(currentBuffer.getOutputStream(), event, codecContext);
+                    }
+                    codec.writeEvent(event, currentBuffer.getOutputStream());
+                    int count = currentBuffer.getEventCount() + 1;
+                    currentBuffer.setEventCount(count);
+
+                    bufferedEventHandles.add(event.getEventHandle());
+                } catch (Exception ex) {
+                    if(sampleException == null) {
+                        sampleException = ex;
+                    }
+                    failedEvents.add(event);
+                }
+
+                flushToLambdaIfNeeded();
+            }
+        } finally {
+            reentrantLock.unlock();
+        }
+
+        if(!failedEvents.isEmpty()) {
+            failedEvents
+                    .stream()
+                    .map(Event::getEventHandle)
+                    .forEach(eventHandle -> eventHandle.release(false));
+            LOG.error("Unable to add {} events to buffer. Dropping these events. Sample exception provided.", failedEvents.size(), sampleException);
+        }
+    }
+
+    private void releaseEventHandles(final boolean result) {
+        for (EventHandle eventHandle : bufferedEventHandles) {
+            eventHandle.release(result);
+        }
+        bufferedEventHandles.clear();
+    }
+
+    private void flushToLambdaIfNeeded() {
+        LOG.trace("Flush to Lambda check: currentBuffer.size={}, currentBuffer.events={}, currentBuffer.duration={}",
+                currentBuffer.getSize(), currentBuffer.getEventCount(), currentBuffer.getDuration());
+        final AtomicReference<String> errorMsgObj = new AtomicReference<>();
+
+        try {
+            if (ThresholdCheck.checkThresholdExceed(currentBuffer, maxEvents, maxBytes, maxCollectionDuration, isBatchEnabled)) {
+                try {
+                    codec.complete(currentBuffer.getOutputStream());
+                    LOG.info("Writing {} to Lambda with {} events and size of {} bytes.",
+                            functionName, currentBuffer.getEventCount(), currentBuffer.getSize());
+                    final boolean isFlushToLambda = retryFlushToLambda(currentBuffer, errorMsgObj);
+
+                    if (isFlushToLambda) {
+                        LOG.info("Successfully flushed to Lambda {}.", functionName);
+                        numberOfRecordsSuccessCounter.increment(currentBuffer.getEventCount());
+                    } else {
+                        LOG.error("Failed to save to Lambda {}", functionName);
+                        numberOfRecordsFailedCounter.increment(currentBuffer.getEventCount());
+                        SdkBytes payload = currentBuffer.getPayload();
+                        dlqPushHandler.perform(pluginSetting,new LambdaSinkFailedDlqData(payload,errorMsgObj.get(),0));
+                    }
+                    //release even if failed
+                    releaseEventHandles(true);
```

Comment:
Should be released with `false` if failed.

Reply:
I think it is considered successful if it is uploaded to dlq?